### ISSUE TITLE
Going to nether disconnects player (permanently, also)

### ISSUE THREAD

**Issue by suresttexas00** [benbaptist/minecraft-wrapper#197]
I discovered another instance of disconnection (maybe this was the problem after all..)  going to the nether causes the same problem disconnect.  I had to turn off nether for now:

```
[12:53:31] [Server thread/INFO]: ry121221 left the game
[12:54:01] [Wrapper.py/INFO] SurestTexas00 executed: /home
[12:54:01] [Server thread/INFO]: Teleported SurestTexas00 to 1591.06077751, 69.0, 463.377406
696
[12:54:01] [Server thread/INFO]: Set SurestTexas00's spawn point to (1591, 70, 463)
[12:54:01] [Server thread/INFO]: Set SurestTexas00's game mode to Survival Mode
[12:54:04] [Wrapper.py/INFO] SurestTexas00 executed: /warps
[12:54:57] [Server thread/INFO]: <ºcSurestTexas00ºr> hey epic.. can you put you stuff away a
nd go to nether so i can re-create the gltich again
[12:55:08] [Server thread/INFO]: <Epic_Skillz_> yes, sure
[12:55:11] [Wrapper.py/INFO] Epic_Skillz_ executed: /home
[12:55:11] [Server thread/INFO]: Teleported Epic_Skillz_ to 6993.30601986, 71.0, 9722.716078
16
[12:55:11] [Server thread/INFO]: Set Epic_Skillz_'s spawn point to (6993, 72, 9722)
[12:55:13] [Server thread/INFO]: <ºcSurestTexas00ºr> once i get the log, ill fix you again
[12:55:34] [Server thread/INFO]: <ºcSurestTexas00ºr> might be 15 minutes after you get kicke
d before i fix it
[12:55:38] [Server thread/INFO]: <ºcSurestTexas00ºr> at most
[12:55:44] [Server thread/INFO]: <Epic_Skillz_> okay
[12:55:56] [Server thread/INFO]: <Epic_Skillz_> here i go
[12:56:02] [Server thread/INFO]: <ºcSurestTexas00ºr> ill assume that if you leave at this po
int, you were kickd
Error in the Server->Client method:
Traceback (most recent call last):
  File "Wrapper\proxy.py", line 908, in handle
    if self.parse(id, original) and self.safe:
  File "Wrapper\proxy.py", line 810, in parse
    chunkColumn += bytearray(self.packet.read_data(16*16*16 * 2)) # packetanisc
  File "Wrapper\proxy.py", line 1143, in read_data
    self.disconnect("Received no data or less data than expected - connection closed")
AttributeError: Packet instance has no attribute 'disconnect'

Last packet IDs (Server->Client) before disconnection:
[('0x8', 34), ('0x44', 41), ('0x3', 17), ('0x30', 94), ('0x2f', 6), ('0x9', 2), ('0x6', 10),
 ('0x1f', 8), ('0x13', 161), ('0x26', 821863)]
[12:56:19] [Server thread/INFO]: Epic_Skillz_ lost connection: TranslatableComponent{key='di
sconnect.genericReason', args=[Internal Exception: java.io.IOException: An existing connecti
on was forcibly closed by the remote host], siblings=[], style=Style{hasParent=false, color=
null, bold=null, italic=null, underlined=null, obfuscated=null, clickEvent=null, hoverEvent=
null, insertion=null}}
[12:56:19] [Server thread/INFO]: Epic_Skillz_ left the game
Failed to grab packet (CLIENT):
Traceback (most recent call last):
  File "Wrapper\proxy.py", line 547, in handle
    id, original = self.packet.grabPacket()
  File "Wrapper\proxy.py", line 945, in grabPacket
    dataLength = self.unpack_varInt()
  File "Wrapper\proxy.py", line 969, in unpack_varInt
    val = struct.unpack('B', self.recv(1))[0]
  File "Wrapper\proxy.py", line 1127, in recv
    d = self.socket.recv(length)
  File "C:\Python27\lib\socket.py", line 170, in _dummy
    raise error(EBADF, 'Bad file descriptor')
error: [Errno 9] Bad file descriptor

[12:56:23] [Wrapper.py/INFO] Epic_Skillz_ logged in (UUID: 8ac1744f-3ad8-474a-a2cb-2cb2664d7
9d3 | IP: 94.227.86.77)
[12:56:24] [Server thread/INFO]: Epic_Skillz_[/127.0.0.1:61111] logged in with entity id 118
205 at (879.6810198794889, 78.0, 1223.5)
[12:56:24] [Server thread/INFO]: Epic_Skillz_ joined the game
Error in the Server->Client method:
Traceback (most recent call last):
  File "Wrapper\proxy.py", line 908, in handle
    if self.parse(id, original) and self.safe:
  File "Wrapper\proxy.py", line 810, in parse
    chunkColumn += bytearray(self.packet.read_data(16*16*16 * 2)) # packetanisc
  File "Wrapper\proxy.py", line 1143, in read_data
    self.disconnect("Received no data or less data than expected - connection closed")
AttributeError: Packet instance has no attribute 'disconnect'

Last packet IDs (Server->Client) before disconnection:
[('0x38', 37), ('0x38', 40), ('0x38', 36), ('0x8', 34), ('0x44', 41), ('0x3', 17), ('0x30',
94), ('0x2f', 6), ('0x2', 320), ('0x26', 821863)]
[12:56:24] [Server thread/INFO]: Epic_Skillz_ lost connection: TranslatableComponent{key='di
sconnect.genericReason', args=[Internal Exception: java.io.IOException: An existing connecti
on was forcibly closed by the remote host], siblings=[], style=Style{hasParent=false, color=
null, bold=null, italic=null, underlined=null, obfuscated=null, clickEvent=null, hoverEvent=
null, insertion=null}}
[12:56:24] [Server thread/INFO]: Epic_Skillz_ left the game
Failed to grab packet (CLIENT):
Traceback (most recent call last):
  File "Wrapper\proxy.py", line 547, in handle
    id, original = self.packet.grabPacket()
  File "Wrapper\proxy.py", line 945, in grabPacket
    dataLength = self.unpack_varInt()
  File "Wrapper\proxy.py", line 969, in unpack_varInt
    val = struct.unpack('B', self.recv(1))[0]
  File "Wrapper\proxy.py", line 1127, in recv
    d = self.socket.recv(length)
  File "C:\Python27\lib\socket.py", line 170, in _dummy
    raise error(EBADF, 'Bad file descriptor')
error: [Errno 9] Bad file descriptor

```

**Comment by benbaptist** [comment]
Hmm

**Comment by benbaptist** [comment]
Proxy mode scares me. MInecraft is horribly quirky with its SMP protocol, and it never makes sense why this stuff happens. I should make sure that there's no extra try:except statements anywhere that might possibly be muddying the errors.

**Comment by kreezxil** [comment]
Could it have something to do with running a bukkit server? I'm running a straight Forge mod server on b1446 and am not getting this no matter how many times I run to and from the Nether.

**Comment by suresttexas00** [comment]
I use an unmodded vanilla client.  currently 1.8.7

**Comment by suresttexas00** [comment]
I probably am using more of the wrapper events than most people, as I have implemented chat filters, protected areas, broadcaster, and various staff and player commands

**Comment by suresttexas00** [comment]
My players positions are also logged every second, etc... so I am processing more packet information and am more likely to see issues.  Also, I am running proxy mode to do all this.

**Comment by suresttexas00** [comment]
hmmm.. why did you think I was running a bukkit server??

**Comment by kreezxil** [comment]
Because the output didn't look standard to me. Don't get mad if you think I
thought you were running bukkit.

**Comment by suresttexas00** [comment]
No anger here.. I was simply curious. The non-standard output is from the wrapper errors.

**Comment by suresttexas00** [comment]
I have colored signs and holograms at spawn (invisible armor stands) too.  I regularly have players question whether my server is vanilla LOL... just never had anyone assume that based on console output before... :D

**Comment by kreezxil** [comment]
Lol, on second look, I don't know why I thought the output reminded me of a bukkit server. Sorry to have caused an aneurysm. :(

However, I'm running FORGE b1446 when I doubled checked and everything was fine. 

Now I have to wonder if the errors aren't in part to running a bleeding edge server based of practically the latest version of Minecraft. However, I'm just some bloke passing through and your error caught my eye.

Oh and I'm using this wrapper too, now. I was a massive CLI script, but I find this to be somewhat much better.

**Comment by suresttexas00** [comment]
It might have been the /warps and /home being executed or my colored name (teams) `<ºcSurestTexas00ºr>`..  supposedly all they are fixing ever since 1.8.4 is "security issues", and it is being released as stable and not a snapshot.  I'm sure I see it more because I have more plugins running.  I opened my nether a day or so ago; i'll solicit player feedback to see if this caused issues still.

**Comment by suresttexas00** [comment]
Another string of disconnections

```
[20:58:29] [Wrapper.py/ERROR]
[20:58:29] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'player.place':
[20:58:29] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:29] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:29] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:29] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 118, in action_right
click
[20:58:29] [Wrapper.py/ERROR]     gamemode = int(player.getGamemode())
[20:58:29] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 85, in getGamemode
[20:58:29] [Wrapper.py/ERROR]     return self.getClient().gamemode
[20:58:29] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'gamemode'
[20:58:29] [Wrapper.py/ERROR]
[20:58:30] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:30] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:30] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:30] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:30] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:30] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:30] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:30] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:30] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:30] [Wrapper.py/ERROR]
[20:58:31] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:31] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:31] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:31] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:31] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:31] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:31] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:31] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:31] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:31] [Wrapper.py/ERROR]
[20:58:32] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:32] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:32] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:32] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:32] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:32] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:32] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:32] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:32] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:32] [Wrapper.py/ERROR]
[20:58:33] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:33] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:33] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:33] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:33] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:33] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:33] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:33] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:33] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:33] [Wrapper.py/ERROR]
[20:58:34] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'player.place':
[20:58:34] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:34] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:34] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:34] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 118, in action_right
click
[20:58:34] [Wrapper.py/ERROR]     gamemode = int(player.getGamemode())
[20:58:34] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 85, in getGamemode
[20:58:34] [Wrapper.py/ERROR]     return self.getClient().gamemode
[20:58:34] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'gamemode'
[20:58:34] [Wrapper.py/ERROR]
[20:58:34] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:34] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:34] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:34] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:34] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:34] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:34] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:34] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:34] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:34] [Wrapper.py/ERROR]
[20:58:35] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:35] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:35] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:35] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:35] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:35] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:35] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:35] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:35] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:35] [Wrapper.py/ERROR]
[20:58:36] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:36] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:36] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:36] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:36] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:36] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:36] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:36] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:36] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:36] [Wrapper.py/ERROR]
[20:58:37] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:37] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:37] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:37] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:37] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:37] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:37] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:37] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:37] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:37] [Wrapper.py/ERROR]
[20:58:38] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:38] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:38] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:38] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:38] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:38] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:38] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:38] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:38] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:38] [Wrapper.py/ERROR]
[20:58:39] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:39] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:39] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:39] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:39] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:39] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:39] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:39] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:39] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'

[20:58:39] [Wrapper.py/ERROR]
[20:58:40] [Wrapper.py/ERROR] Plugin 'com.suresttexas00.plugins.logger' errored out when exe
cuting callback event 'timer.second':
[20:58:40] [Wrapper.py/ERROR] Traceback (most recent call last):
[20:58:40] [Wrapper.py/ERROR]   File "Wrapper\events.py", line 33, in callEvent
[20:58:40] [Wrapper.py/ERROR]     result = self.events[pluginID][event](payload)
[20:58:40] [Wrapper.py/ERROR]   File "wrapper-plugins\regions.py", line 213, in onSecond
[20:58:40] [Wrapper.py/ERROR]     dim = player.getDimension()
[20:58:40] [Wrapper.py/ERROR]   File "Wrapper\api\player.py", line 88, in getDimension
[20:58:40] [Wrapper.py/ERROR]     return self.getClient().dimension
[20:58:40] [Wrapper.py/ERROR] AttributeError: 'NoneType' object has no attribute 'dimension'
```

Offending plugin code;
```
    def onSecond(self, payload):
        for players in self.api.minecraft.getPlayers():
            player = self.api.minecraft.getPlayer(players)
            playeruuid = str(player.uuid)
            dim = player.getDimension()
            pos = player.getPosition()
...
```

**Comment by suresttexas00** [comment]
D':

**Comment by suresttexas00** [comment]
It is my suspicion that the player list changes while the for loop is processing the records.. resulting in errors for the getPlayer method.  I'm going to try beefing up my code and stop assuming wrapper will return solid data:
```
    def onSecond(self, payload):
        for players in self.api.minecraft.getPlayers():
            player = self.api.minecraft.getPlayer(players)
            playeruuid = str(player.uuid)
            dim = None
            try:
                dim = player.getDimension()
                pos = player.getPosition()
            except:
                pass
            if dim is None:
                continue
```

**Comment by benbaptist** [comment]
I managed to almost reproduce this issue. http://s.benbaptist.com/V3AR8

I have a Platform Power map world that seems to do this a lot. Almost every time I connect for the first time since booting Wrapper.py and the server, it gives me that error seen above. However, this is not permanent - I can connect the second time usually. What happens differently from the first connect over the second connect since a server has been booted? 

For the record, Platform Power is a minigame, and I have seen the permanent issue happen as well on Platform Power rounds where no matter how many times you connect, you always fail/crash.

**Comment by benbaptist** [comment]
Interesting. I can tell that it appears to be consistent when the server sends a 0x1a packet. I figured this out by sleeping before each send by 0.1 seconds to make it easier to see each packet being sent in real time. Then, I printed out each packet ID being sent to the client. It was consistently after the 0x1a packet that the client died. 

0x1a is the "Entity Status" packet. It was sending 0 for both the entity ID and status value. Oh, and now that I look at it, I see that I was receiving the entity ID as a varint when it is an integer. That might've been causing an issue. The question is, though - why did this only happen on the first connect? I'm really confused.
